beacon-chain/blockchain/chain_info.go

```diff
@@ -329,24 +341,25 @@ func (s *Service) IsOptimisticForRoot(ctx context.Context, root [32]byte) (bool,
 	if err != protoarray.ErrUnknownNodeRoot && err != doublylinkedtree.ErrNilNode {
 		return false, err
 	}
+	// if the requested root is the headroot we should treat the
+	// node as optimistic. This can happen if we pruned INVALID
+	// nodes and no viable head is available.
+	headRoot, err := s.HeadRoot(ctx)
+	if err != nil {
+		return true, err
+	}
+	if bytes.Equal(headRoot, root[:]) {
+		return true, nil
+	}
+
 	ss, err := s.cfg.BeaconDB.StateSummary(ctx, root)
 	if err != nil {
 		return false, err
 	}
+
 	if ss == nil {
```

Comment:
Why not do this so we don't have to worry about `s?HeadRoot` at all
```suggestion
	if ss == nil {
	    	if s.cfg.ForkChoiceStore.AllTipsAreInvalid() {
		return true, nil
	        }
```

Reply:
because forkchoice's `AllTipsAreInvalid` may fail if there was no call to `Head()` between the prunning and the call here.

Reply:
But I agree that it's such a long shot that someone requests by root (instead of just `IsOptimistic`) exactly at this same time. I'll let you decide, if you want me to switch this to avoid the comparison I'm happy with it... this is not really a hot path

Reply:
I'm just thinking future proof. What if we were to remove `HeadRoot` from blockchain's `chaininfo` one day and default everything to the fork choice store's head?

Reply:
Ah but the problem of doing the suggestion is that someone may actually be requesting by root an old block, we should reply correctly instead of "true" if forkchoice is in an incoherent state